internal/controller/bindings/forwarder_controller.go

```diff
@@ -132,40 +241,27 @@ func (r *ForwarderReconciler) statusID(epb *bindingsv1alpha1.BoundEndpoint) stri
 	return fmt.Sprintf("%s/%s", epb.Namespace, epb.Name)
 }
 
-// exampleConnectionHandler is a simple example of a connection handler that echos back each line
-// it reads from the client. It also sends a welcome message to the client.
-func exampleConnectionHandler(epb *bindingsv1alpha1.BoundEndpoint) bindingsdriver.ConnectionHandler {
-	return func(conn net.Conn) error {
-		defer conn.Close()
-		_, err := conn.Write([]byte(
-			fmt.Sprintf(
-				`
-Hello from the ngrok-operator bindings-forwarder
-
-You are connected to my port: %d
-The endpoint binding is: %s/%s
-
-Type anything and hit enter to see it echoed back
-Type an empty line to close the connection
-
->`, epb.Spec.Port, epb.Namespace, epb.Name,
-			),
-		))
-		if err != nil {
-			return err
-		}
-		scanner := bufio.NewScanner(conn)
-		for scanner.Scan() {
-			line := scanner.Text()
-			if line == "" { // empty line means the client is done
-				break
+func joinConnections(log logr.Logger, conn1, conn2 net.Conn) error {
+	var g errgroup.Group
+	g.Go(func() error {
+		defer func() {
+			if err := conn1.Close(); err != nil {
+				log.Error(err, "failed closing connection to destination: %v", err)
 			}
+		}()
 
-			_, err := conn.Write([]byte(line + "\n> "))
-			if err != nil {
-				return err
+		_, err := io.Copy(conn1, conn2)
+		return err
+	})
+	g.Go(func() error {
+		defer func() {
+			if err := conn2.Close(); err != nil {
+				log.Error(err, "failed closing connection to client: %v", err)
 			}
-		}
-		return nil
-	}
+		}()
+
+		_, err := io.Copy(conn2, conn1)
+		return err
+	})
+	return g.Wait()
```

Comment:
Q. If the connection is interrupted or closed unexpectedly, what is the behaviour of the pod forwarder to re-dial?

Reply:
The forwarder only dials on the initial connection. If the upstream connection to mux is closed, it will disconnect the in-cluster client and they will have to re-dial.

Reply:
Who has to re-dial? This would be the forwarder needing to re-open the tcp connection, no?

Reply:
The client of bindings-forwarder would need to try to re-connect to in-cluster service. If the connection to the ingress endpoint is interrupted or closed, the connection to the in-cluster client is no longer usable.

Reply:
Chatted through this in Slack